### PR TITLE
EXIF data for ios camera

### PR DESCRIPTION
**Pull request by PeterPetrik**
Implementation of EXIF for iOS using Camera. 
GPS data used from QgsQuickPositionKit and QCompass.

https://user-images.githubusercontent.com/6735606/114743968-7c657a80-9d4d-11eb-939f-a9c79a3d2d94.MP4

### CODE REVIEW

File: app/ios/iosinterface.mm [hunk 1]
```diff
@@ -168,48 +170,72 @@ -( void )showImagePicker:( int )sourceType : ( IOSImagePicker * )handler
...
         {
-          QString imagePathData( [imagePath UTF8String] );
-          data["imagePath"] = imagePathData;
+          capturedImage = info[UIImagePickerControllerOriginalImage];
         }
 
-        if ( delegate->handler )
-        {
+        // 2. Get your metadata (includes the EXIF data).
+        NSDictionary *metadata = [info objectForKey:UIImagePickerControllerMediaMetadata];
```
Comment (sklencar) [inline app/ios/iosinterface.mm:189]: Doesnt contain GPS data.
Example:
```
{
    DPIHeight = 72;
    DPIWidth = 72;
    Orientation = 6;
    kCGImageDestinationICCProfile = {length = 548, bytes = 0x00000224 6170706c 04000000 6d6e7472 ... 000003dc 0000c06e };
    "{Exif}" =     {
        ApertureValue = "1.6959938131099";
 
       BrightnessValue = "4.355585555348755";
        ColorSpace = 65535;
        CompositeImage = 2;
        DateTimeDigitized = "2021:04:13 12:16:55";
        DateTimeOriginal = "2021:04:13 12:16:55";
        ExifVersion = 0232;
        ExposureBia
sValue = 0;
        ExposureMode = 0;
        ExposureProgram = 2;
        ExposureTime = "0.008264462809917356";
        FNumber = "1.8";
        Flash = 24;
        FocalLenIn35mmFilm = 26;
        FocalLength = "4.25";
        ISOSpeedRatings = 
        (
            80
        );
        LensMake = Apple;
        LensModel = "iPhone 11 back dual wide camera 4.25mm f/1.8";
        LensSpecification =         (
            "1.539999961853027",
            "4.25",
            "1.8",
       
     "2.400000095367432"
        );
        MeteringMode = 5;
        OffsetTime = "+02:00";
        OffsetTimeDigitized = "+02:00";
        OffsetTimeOriginal = "+02:00";
        PixelXDimension = 4032;
        PixelYDimension = 3024;
        Scen
eType = 1;
        SensingMethod = 2;
        ShutterSpeedValue = "6.924541302430206";
        SubjectArea =         (
            2003,
            1505,
            2112,
            1264
        );
        SubsecTimeDigitized = 481;
        Su
bsecTimeOriginal = 481;
        WhiteBalance = 0;
    };
    "{MakerApple}" =     {
        1 = 11;
        12 =         (
            "102.234375",
            "97.75"
        );
        13 = 5;
        14 = 0;
        15 = 2;
        16 = 3;

        2 = {length = 512, bytes = 0x4c014201 2f011401 f900df00 c800a900 ... 08000600 04000600 };
        20 = 10;
        23 = 39321600;
        25 = 2;
        3 =         {
            epoch = 0;
            flags = 1;
            timescale = 10
00000000;
            value = 406211674554041;
        };
        31 = 0;
        32 = "251D673E-C792-4D82-9BC3-4C3D17E9800D";
        33 = 0;
        35 =         (
            7,
            268435556
        );
        37 = 138;
        38 = 
3;
        39 = "41.97124862670898";
        4 = 0;
        40 = 2;
        43 = "0F24CEBA-794D-40A9-BFBF-F242EBA7B954";
        5 = 194;
        6 = 169;
        7 = 1;
        8 =         (
            "-0.002626766683533788",
            "-0.5
14091968536377",
            "-0.6946022510528564"
        );
    };
    "{TIFF}" =     {
        DateTime = "2021:04:13 12:16:55";
        Make = Apple;
        Model = "iPhone 11";
        ResolutionUnit = 2;
        Software = "13.6";
        
XResolution = 72;
        YResolution = 72;
    };
}
```

File: app/ios/iosinterface.mm [hunk 2]
```diff
@@ -122,6 +66,105 @@ static QImage fromUIImage( UIImage *image )
   return result;
 }
 
++( void )handleCameraPhoto:( NSDictionary * )info:( NSString * )imagePath:( QgsQuickPositionKit * )positionKit:( PositionDirection * )compass
+{
+  qDebug() << "Call method!!";
```
Comment (PeterPetrik) [inline app/ios/iosinterface.mm:71]: remove

File: app/ios/iosinterface.mm [hunk 3]
```diff
@@ -122,6 +66,105 @@ static QImage fromUIImage( UIImage *image )
...
+  {
+    err = "failed to create image destination.";
+  }
+  else
+  {
+    // 6. Save the image
+    CGImageDestinationAddImage( imageDestination, capturedImage.CGImage, ( __bridge CFDictionaryRef )mutableMetadata );
+    if ( CGImageDestinationFinalize( imageDestination ) == NO )
+    {
+      err = "failed to finalize the image.";
```
Comment (PeterPetrik) [inline app/ios/iosinterface.mm:105]: not used? change to qDebug() ?

File: app/ios/iosinterface.mm [hunk 4]
```diff
@@ -122,6 +66,105 @@ static QImage fromUIImage( UIImage *image )
...
+    [mutableMetadata setObject:mGpsData forKey:( NSString * )kCGImagePropertyGPSDictionary];
+
+  // 4. Set your compression quuality (0.0 to 1.0).
+  [mutableMetadata setObject:@( 1.0 ) forKey:( __bridge NSString * )kCGImageDestinationLossyCompressionQuality];
+
+  // 5. Create an image destination.
+  CGImageDestinationRef imageDestination = CGImageDestinationCreateWithURL( ( __bridge CFURLRef )outputURL, kUTTypeJPEG, 1, NULL );
+  if ( imageDestination == NULL )
+  {
+    err = "failed to create image destination.";
```
Comment (PeterPetrik) [inline app/ios/iosinterface.mm:97]: not used? change to qDebug() ?

File: app/ios/iosinterface.mm [hunk 5]
```diff
@@ -159,98 +202,28 @@ -( void )showImagePicker:( int )sourceType : ( IOSImagePicker * )handler
     static IOSViewDelegate *delegate = nullptr;
     delegate = [[IOSViewDelegate alloc] initWithHandler:handler];
 
+    // TODO use NSNotificationCenter::post and userInfo to pass GPSData dict instead of global variable.
```
Comment (PeterPetrik) [inline app/ios/iosinterface.mm:205]: remove todo, this is OK

File: app/ios/iosutils.cpp [hunk 6]
```diff
@@ -29,12 +29,12 @@ bool IosUtils::isIos() const
...
 void IosUtils::callImagePicker( const QString &targetPath )
 {
-  mImagePicker->showImagePicker( 0, targetPath, nullptr );
+  mImagePicker->showImagePicker( 0, targetPath, nullptr, nullptr );
 }
 
 void IosUtils::callCamera( const QString &targetPath )
 {
-  mImagePicker->showImagePicker( 1, targetPath, mPositionKit );
+  mImagePicker->showImagePicker( 1, targetPath, mPositionKit, mCompass );
```
Comment (PeterPetrik) [inline app/ios/iosutils.cpp:37]: remove showImagePicker function and create 2 native functions, one for Camera, one for gallery

File: app/ios/iosutils.h [hunk 7]
```diff
@@ -39,11 +41,14 @@ class IosUtils: public QObject
...
     void imageSelected( const QString &imagePath );
     void positionKitChanged();
+    void compassChanged();
 
   private:
 
     IOSImagePicker *mImagePicker = nullptr;
     QgsQuickPositionKit *mPositionKit = nullptr;
+    // TODO
+    PositionDirection *mCompass = nullptr;
```
Comment (PeterPetrik) [inline app/ios/iosutils.h:51]: in the whole PR, mCompass should be new class "Compass". I assume this is still WIP?

**Comment by saberraz** [comment]
Tested in iOS and works as expected 👏👏👏

**Pull request merged**